File: ios/runanywhere-sdks-latest/sdk/runanywhere-flutter/packages/runanywhere/lib/public/types/tool_calling_types.dart
/// Tool Calling Types for RunAnywhere SDK
///
/// Type definitions for tool calling (function calling) functionality.
/// Allows LLMs to request external actions (API calls, device functions, etc.)
///
/// Mirrors Swift SDK's ToolCallingTypes.swift
library tool_calling_types;

import 'dart:convert';

// =============================================================================
// TOOL CALL FORMAT NAMES
// =============================================================================

/// Constants for tool call format names.
///
/// The format logic is handled in C++ commons (single source of truth).
/// Mirrors Swift SDK's ToolCallFormatName enum.
abstract class ToolCallFormatName {
  /// JSON format: `<tool_call>{"tool":"name","arguments":{...}}</tool_call>`
  /// Use for most general-purpose models (Llama, Qwen, Mistral, etc.)
  static const String defaultFormat = 'default';

  /// Liquid AI format: `<|tool_call_start|>[func(args)]<|tool_call_end|>`
  /// Use for LFM2-Tool models
  static const String lfm2 = 'lfm2';
}

// =============================================================================
// TOOL VALUE - Type-safe JSON representation
// =============================================================================

/// A type-safe representation of JSON values for tool arguments and results.
/// Avoids using `dynamic` while supporting all JSON types.
sealed class ToolValue {
  const ToolValue();

  // Convenience value extraction
  String? get stringValue => this is StringToolValue ? (this as StringToolValue).value : null;
  double? get numberValue => this is NumberToolValue ? (this as NumberToolValue).value : null;
  int? get intValue => numberValue?.toInt();
  bool? get boolValue => this is BoolToolValue ? (this as BoolToolValue).value : null;
  List<ToolValue>? get arrayValue =>
      this is ArrayToolValue ? (this as ArrayToolValue).value : null;
  Map<String, ToolValue>? get objectValue =>
      this is ObjectToolValue ? (this as ObjectToolValue).value : null;
  bool get isNull => this is NullToolValue;

  /// Convert to JSON-compatible dynamic value
  dynamic toJson() => switch (this) {
        StringToolValue(value: var v) => v,
        NumberToolValue(value: var v) => v,
        BoolToolValue(value: var v) => v,
        ArrayToolValue(value: var v) => v.map((e) => e.toJson()).toList(),
        ObjectToolValue(value: var v) => v.map((k, val) => MapEntry(k, val.toJson())),
        NullToolValue() => null,
      };

  /// Create from any JSON-compatible value
  static ToolValue from(dynamic value) => switch (value) {
        null => const NullToolValue(),
        String s => StringToolValue(s),
        num n => NumberToolValue(n.toDouble()),
        bool b => BoolToolValue(b),
        List l => ArrayToolValue(l.map(from).toList()),
        Map m => ObjectToolValue(m.map((k, v) => MapEntry(k.toString(), from(v)))),
        _ => StringToolValue(value.toString()),
      };
}

class StringToolValue extends ToolValue {
  final String value;
  const StringToolValue(this.value);
}

class NumberToolValue extends ToolValue {
  final double value;
  const NumberToolValue(this.value);
}

class BoolToolValue extends ToolValue {
  final bool value;
  const BoolToolValue(this.value);
}

class ArrayToolValue extends ToolValue {
  final List<ToolValue> value;
  const ArrayToolValue(this.value);
}

class ObjectToolValue extends ToolValue {
  final Map<String, ToolValue> value;
  const ObjectToolValue(this.value);
}

class NullToolValue extends ToolValue {
  const NullToolValue();
}

// =============================================================================
// PARAMETER TYPES
// =============================================================================

/// Supported parameter types for tool arguments
enum ToolParameterType {
  string('string'),
  number('number'),
  boolean('boolean'),
  object('object'),
  array('array');

  final String value;
  const ToolParameterType(this.value);

  static ToolParameterType fromString(String value) => switch (value.toLowerCase()) {
        'string' => ToolParameterType.string,
        'number' => ToolParameterType.number,
        'boolean' => ToolParameterType.boolean,
        'object' => ToolParameterType.object,
        'array' => ToolParameterType.array,
        _ => ToolParameterType.string,
      };
}

/// A single parameter definition for a tool
class ToolParameter {
  /// Parameter name
  final String name;

  /// Data type of the parameter
  final ToolParameterType type;

  /// Human-readable description
  final String description;

  /// Whether this parameter is required
  final bool required;

  /// Allowed values (for enum-like parameters)
  final List<String>? enumValues;

  const ToolParameter({
    required this.name,
    required this.type,
    required this.description,
    this.required = true,
    this.enumValues,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'type': type.value,
        'description': description,
        'required': required,
        if (enumValues != null) 'enumValues': enumValues,
      };
}

// =============================================================================
// TOOL DEFINITION TYPES
// =============================================================================

/// Definition of a tool that the LLM can use
class ToolDefinition {
  /// Unique name of the tool (e.g., "get_weather")
  final String name;

  /// Human-readable description of what the tool does
  final String description;

  /// Parameters the tool accepts
  final List<ToolParameter> parameters;

  /// Category for organizing tools (optional)
  final String? category;

  const ToolDefinition({
    required this.name,
    required this.description,
    required this.parameters,
    this.category,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'description': description,
        'parameters': parameters.map((p) => p.toJson()).toList(),
        if (category != null) 'category': category,
      };
}

// =============================================================================
// TOOL CALL TYPES (LLM requesting to use a tool)
// =============================================================================

/// A request from the LLM to execute a tool
class ToolCall {
  /// Name of the tool to execute
  final String toolName;

  /// Arguments to pass to the tool
  final Map<String, ToolValue> arguments;

  /// Unique ID for this tool call (for tracking)
  final String? callId;

  const ToolCall({
    required this.toolName,
    required this.arguments,
    this.callId,
  });

  /// Get a string argument by name
  String? getString(String key) => arguments[key]?.stringValue;

  /// Get a number argument by name
  double? getNumber(String key) => arguments[key]?.numberValue;

  /// Get a bool argument by name
  bool? getBool(String key) => arguments[key]?.boolValue;
}

// =============================================================================
// TOOL RESULT TYPES (Result after execution)
// =============================================================================

/// Result of executing a tool
class ToolResult {
  /// Name of the tool that was executed
  final String toolName;

  /// Whether execution was successful
  final bool success;

  /// Result data (if successful)
  final Map<String, ToolValue>? result;

  /// Error message (if failed)
  final String? error;

  /// The original call ID (for tracking)
  final String? callId;

  const ToolResult({
    required this.toolName,
    required this.success,
    this.result,
    this.error,
    this.callId,
  });

  Map<String, dynamic> toJson() => {
        'toolName': toolName,
        'success': success,
        if (result != null) 'result': result!.map((k, v) => MapEntry(k, v.toJson())),
        if (error != null) 'error': error,
        if (callId != null) 'callId': callId,
      };
}

// =============================================================================
// TOOL EXECUTOR TYPES
// =============================================================================

/// Function type for tool executors.
/// Takes arguments as strongly-typed ToolValue map, returns result map.
typedef ToolExecutor = Future<Map<String, ToolValue>> Function(Map<String, ToolValue> args);

// =============================================================================
// TOOL CALLING OPTIONS
// =============================================================================

/// Options for tool-enabled generation
class ToolCallingOptions {
  /// Available tools for this generation (if not provided, uses registered tools)
  final List<ToolDefinition>? tools;

  /// Maximum number of tool calls allowed in one conversation turn (default: 5)
  final int maxToolCalls;

  /// Whether to automatically execute tools or return them for manual execution (default: true)
  final bool autoExecute;

  /// Temperature for generation
  final double? temperature;

  /// Maximum tokens to generate
  final int? maxTokens;

  /// System prompt to use (will be merged with tool instructions by default)
  final String? systemPrompt;

  /// If true, replaces the system prompt entirely instead of appending tool instructions
  final bool replaceSystemPrompt;

  /// If true, keeps tool definitions available after the first tool call
  final bool keepToolsAvailable;

  /// Tool calling format name (e.g., "default", "lfm2")
  /// Different models are trained on different tool calling formats.
  /// - "default": Standard JSON format for general-purpose models
  /// - "lfm2": Pythonic format for LFM2-Tool models
  final String formatName;

  const ToolCallingOptions({
    this.tools,
    this.maxToolCalls = 5,
    this.autoExecute = true,
    this.temperature,
    this.maxTokens,
    this.systemPrompt,
    this.replaceSystemPrompt = false,
    this.keepToolsAvailable = false,
    this.formatName = ToolCallFormatName.defaultFormat,
  });
}

// =============================================================================
// TOOL CALLING RESULT TYPES
// =============================================================================

/// Result of a generation that may include tool calls
class ToolCallingResult {
  /// The final text response
  final String text;

  /// Any tool calls the LLM made
  final List<ToolCall> toolCalls;

  /// Results of executed tools (if autoExecute was true)
  final List<ToolResult> toolResults;

  /// Whether the response is complete or waiting for tool results
  final bool isComplete;

  /// Conversation ID for continuing with tool results
  final String? conversationId;

  const ToolCallingResult({
    required this.text,
    required this.toolCalls,
    required this.toolResults,
    required this.isComplete,
    this.conversationId,
  });
}

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================

/// Serialize tools to JSON string
String toolsToJson(List<ToolDefinition> tools) {
  return jsonEncode(tools.map((t) => t.toJson()).toList());
}

/// Convert Map<String, ToolValue> to JSON string
String toolResultToJsonString(Map<String, ToolValue> result) {
  return jsonEncode(result.map((k, v) => MapEntry(k, v.toJson())));
}

/// Convert Map<String, dynamic> to Map<String, ToolValue>
Map<String, ToolValue> dynamicMapToToolValueMap(Map<String, dynamic> map) {
  return map.map((k, v) => MapEntry(k, ToolValue.from(v)));
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/public/extensions/LLM/ToolCallingTypes.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Type definitions for Tool Calling functionality.
 * Allows LLMs to request external actions (API calls, device functions, etc.)
 *
 * Mirrors Swift SDK's ToolCallingTypes.swift
 */

package com.runanywhere.sdk.public.extensions.LLM

import kotlinx.serialization.Serializable

// =============================================================================
// TOOL VALUE - Type-safe JSON representation
// =============================================================================

/**
 * A type-safe representation of JSON values for tool arguments and results.
 * Avoids using `Any` while supporting all JSON types.
 */
@Serializable
sealed class ToolValue {
    @Serializable
    data class StringValue(
        val value: String,
    ) : ToolValue()

    @Serializable
    data class NumberValue(
        val value: Double,
    ) : ToolValue()

    @Serializable
    data class BoolValue(
        val value: Boolean,
    ) : ToolValue()

    @Serializable
    data class ArrayValue(
        val value: List<ToolValue>,
    ) : ToolValue()

    @Serializable
    data class ObjectValue(
        val value: Map<String, ToolValue>,
    ) : ToolValue()

    @Serializable
    object NullValue : ToolValue()

    // Convenience value extraction
    val stringValue: String? get() = (this as? StringValue)?.value
    val numberValue: Double? get() = (this as? NumberValue)?.value
    val intValue: Int? get() = numberValue?.toInt()
    val boolValue: Boolean? get() = (this as? BoolValue)?.value
    val arrayValue: List<ToolValue>? get() = (this as? ArrayValue)?.value
    val objectValue: Map<String, ToolValue>? get() = (this as? ObjectValue)?.value
    val isNull: Boolean get() = this is NullValue

    companion object {
        fun string(value: String) = StringValue(value)

        fun number(value: Double) = NumberValue(value)

        fun number(value: Int) = NumberValue(value.toDouble())

        fun bool(value: Boolean) = BoolValue(value)

        fun array(value: List<ToolValue>) = ArrayValue(value)

        fun obj(value: Map<String, ToolValue>) = ObjectValue(value)

        fun nullValue() = NullValue

        /**
         * Convert Any to ToolValue
         */
        fun from(value: Any?): ToolValue =
            when (value) {
                null -> NullValue
                is String -> StringValue(value)
                is Number -> NumberValue(value.toDouble())
                is Boolean -> BoolValue(value)
                is List<*> -> ArrayValue(value.map { from(it) })
                is Map<*, *> ->
                    ObjectValue(
                        value.entries.associate { (k, v) ->
                            k.toString() to from(v)
                        },
                    )
                else -> StringValue(value.toString())
            }
    }
}

// =============================================================================
// PARAMETER TYPES
// =============================================================================

/**
 * Supported parameter types for tool arguments
 */
enum class ToolParameterType(
    val value: String,
) {
    STRING("string"),
    NUMBER("number"),
    BOOLEAN("boolean"),
    OBJECT("object"),
    ARRAY("array"),
    ;

    companion object {
        fun fromString(value: String): ToolParameterType =
            when (value.lowercase()) {
                "string" -> STRING
                "number" -> NUMBER
                "boolean" -> BOOLEAN
                "object" -> OBJECT
                "array" -> ARRAY
                else -> STRING
            }
    }
}

/**
 * A single parameter definition for a tool
 */
data class ToolParameter(
    /** Parameter name */
    val name: String,
    /** Data type of the parameter */
    val type: ToolParameterType,
    /** Human-readable description */
    val description: String,
    /** Whether this parameter is required */
    val required: Boolean = true,
    /** Allowed values (for enum-like parameters) */
    val enumValues: List<String>? = null,
)

// =============================================================================
// TOOL DEFINITION TYPES
// =============================================================================

/**
 * Definition of a tool that the LLM can use
 */
data class ToolDefinition(
    /** Unique name of the tool (e.g., "get_weather") */
    val name: String,
    /** Human-readable description of what the tool does */
    val description: String,
    /** Parameters the tool accepts */
    val parameters: List<ToolParameter>,
    /** Category for organizing tools (optional) */
    val category: String? = null,
)

// =============================================================================
// TOOL CALL TYPES (LLM requesting to use a tool)
// =============================================================================

/**
 * A request from the LLM to execute a tool
 */
data class ToolCall(
    /** Name of the tool to execute */
    val toolName: String,
    /** Arguments to pass to the tool */
    val arguments: Map<String, ToolValue>,
    /** Unique ID for this tool call (for tracking) */
    val callId: String? = null,
) {
    /** Get a string argument by name */
    fun getString(key: String): String? = arguments[key]?.stringValue

    /** Get a number argument by name */
    fun getNumber(key: String): Double? = arguments[key]?.numberValue

    /** Get a bool argument by name */
    fun getBool(key: String): Boolean? = arguments[key]?.boolValue
}

// =============================================================================
// TOOL RESULT TYPES (Result after execution)
// =============================================================================

/**
 * Result of executing a tool
 */
data class ToolResult(
    /** Name of the tool that was executed */
    val toolName: String,
    /** Whether execution was successful */
    val success: Boolean,
    /** Result data (if successful) */
    val result: Map<String, ToolValue>? = null,
    /** Error message (if failed) */
    val error: String? = null,
    /** The original call ID (for tracking) */
    val callId: String? = null,
)

// =============================================================================
// TOOL EXECUTOR TYPES
// =============================================================================

/**
 * Function type for tool executors.
 * Takes arguments as strongly-typed ToolValue map, returns result map.
 */
typealias ToolExecutor = suspend (Map<String, ToolValue>) -> Map<String, ToolValue>

/**
 * A registered tool with its definition and executor
 */
internal data class RegisteredTool(
    val definition: ToolDefinition,
    val executor: ToolExecutor,
)

// =============================================================================
// TOOL CALL FORMAT NAMES
// =============================================================================

/**
 * Format names for tool calling output (internal string constants).
 * Used for C++ bridge communication.
 */
internal object ToolCallFormatName {
    const val DEFAULT = "default"
    const val LFM2 = "lfm2"
}

/**
 * Tool calling format types.
 * Each format specifies how tool calls are formatted in the LLM prompt.
 *
 * The format logic is handled in C++ commons (single source of truth).
 */
sealed class ToolCallFormat {
    /**
     * Default format using XML-style tags.
     * JSON format: `<tool_call>{"tool":"name","arguments":{...}}</tool_call>`
     * Use for most general-purpose models (Llama, Qwen, Mistral, etc.)
     */
    data object Default : ToolCallFormat()

    /**
     * LFM2 format for Liquid AI models.
     * Liquid AI format: `<|tool_call_start|>[func(args)]<|tool_call_end|>`
     * Use for LFM2-Tool models.
     */
    data object LFM2 : ToolCallFormat()

    /** Get the string representation for C++ bridge */
    fun toFormatName(): String =
        when (this) {
            is Default -> ToolCallFormatName.DEFAULT
            is LFM2 -> ToolCallFormatName.LFM2
        }

    companion object {
        /** Convert from format name string (for deserialization) */
        fun fromFormatName(name: String?): ToolCallFormat =
            when (name) {
                ToolCallFormatName.LFM2 -> LFM2
                else -> Default
            }
    }
}

// =============================================================================
// TOOL CALLING OPTIONS
// =============================================================================

/**
 * Options for tool-enabled generation
 */
data class ToolCallingOptions(
    /** Available tools for this generation (if not provided, uses registered tools) */
    val tools: List<ToolDefinition>? = null,
    /** Maximum number of tool calls allowed in one conversation turn (default: 5) */
    val maxToolCalls: Int = 5,
    /** Whether to automatically execute tools or return them for manual execution (default: true) */
    val autoExecute: Boolean = true,
    /** Temperature for generation */
    val temperature: Float? = null,
    /** Maximum tokens to generate */
    val maxTokens: Int? = null,
    /** System prompt to use (will be merged with tool instructions by default) */
    val systemPrompt: String? = null,
    /** If true, replaces the system prompt entirely instead of appending tool instructions */
    val replaceSystemPrompt: Boolean = false,
    /** If true, keeps tool definitions available after the first tool call */
    val keepToolsAvailable: Boolean = false,
    /**
     * Format for tool calls.
     * Use [ToolCallFormat.LFM2] for LFM2-Tool models (Liquid AI).
     * Default: [ToolCallFormat.Default] which uses JSON-based format suitable for most models.
     */
    val format: ToolCallFormat = ToolCallFormat.Default,
)

// =============================================================================
// TOOL CALLING RESULT TYPES
// =============================================================================

/**
 * Result of a generation that may include tool calls
 */
data class ToolCallingResult(
    /** The final text response */
    val text: String,
    /** Any tool calls the LLM made */
    val toolCalls: List<ToolCall>,
    /** Results of executed tools (if autoExecute was true) */
    val toolResults: List<ToolResult>,
    /** Whether the response is complete or waiting for tool results */
    val isComplete: Boolean,
    /** Conversation ID for continuing with tool results */
    val conversationId: String? = null,
)
